src/apps/relay/windows/ServiceInstaller.cpp

```diff
@@ -0,0 +1,162 @@
+                               SC_MANAGER_CONNECT | SC_MANAGER_CREATE_SERVICE);
+  if (NULL == schSCManager) {
+    printf("OpenSCManager failed (%lu)\n", GetLastError());
+    return -2;
+  }
+
+  // Install the service into SCM by calling CreateService
+  schService = CreateService(schSCManager,              // SCManager database
+                             pszServiceName,            // Name of service
+                             pszDisplayName,            // Name to display
+                             SERVICE_QUERY_STATUS,      // Desired access
+                             SERVICE_WIN32_OWN_PROCESS, // Service type
+                             dwStartType,               // Service start type
+                             SERVICE_ERROR_NORMAL,      // Error control type
+                             szPath,                    // Service's binary
+                             NULL,                      // No load ordering group
+                             NULL,                      // No tag identifier
+                             pszDependencies,           // Dependencies
+                             pszAccount,                // Service running account
+                             pszPassword                // Password of the account
+  );
+  if (schService == NULL) {
+    printf("CreateService failed (%lu)\n", GetLastError());
+    CloseServiceHandle(schSCManager);
+    return -3;
+  } else
+    printf("Service installed successfully\n");
+
+  // Centralized cleanup for all allocated resources.
+  CloseServiceHandle(schService);
+  CloseServiceHandle(schSCManager);
+  return 0;
+}
+
+/*!
+ * \brief Uninstall service
+ *
+ * \details Stop and remove the service from the local service control
+ *          manager database.
+ *
+ * \param pszServiceName - the name of the service to be removed.
+ *
+ * \note If the function fails to uninstall the service, it prints the
+ *   error in the standard output stream for users to diagnose the problem.
+ *
+ * \see https://learn.microsoft.com/windows/win32/services/deleting-a-service
+ */
+int UninstallService(LPTSTR pszServiceName) {
+  SC_HANDLE schSCManager = NULL;
+  SC_HANDLE schService = NULL;
+  SERVICE_STATUS ssSvcStatus = {};
+
+  // Open the local default service control manager database
+  schSCManager = OpenSCManager(NULL, NULL, SC_MANAGER_CONNECT);
+  if (schSCManager == NULL) {
+    printf("OpenSCManager failed (%lu)\n", GetLastError());
+    return -1;
+  }
+
+  // Open the service with delete, stop, and query status permissions
+  schService = OpenService(schSCManager, pszServiceName, SERVICE_STOP | SERVICE_QUERY_STATUS | DELETE);
+  if (schService == NULL) {
+    printf("OpenService failed (%lu)\n", GetLastError());
+    CloseServiceHandle(schSCManager);
+    return -2;
+  }
+
+  // Try to stop the service
+  if (ControlService(schService, SERVICE_CONTROL_STOP, &ssSvcStatus)) {
+    _tprintf(_T("Stopping %s."), pszServiceName);
+    Sleep(1000);
```

Comment:
Imho sleeping for a full second seems pretty long. 

I'd change this to a tenth of a second, and i wouldn't bother waiting prior to checking the status, personally.

Reply:
It's not long, you're too impatient  😄

Reply:
It's pretty long when you're service might take tens of seconds to shut down due to the additional sleeps?